### PR TITLE
impl #388, support slot props

### PR DESCRIPTION
**Pull request by Yuyz0112**
Implementation in this PR follows the design described in #388.

Currently, this status is ready for review but not ready for merge because it only migrates two components to the new slot spec.

Also, this implementation introduces a breaking change to the components spec, so we need a major release if we are going to merge this.

### CODE REVIEW

**Comment by Yuyz0112** [comment]
CI will not pass before we migrate all the components.

**Comment by tanbowensg** [comment]
I think we can introduce this pr next tuesday.

**Comment by tanbowensg** [comment]
We need allow user to edit `ifCondition` in editor.

**Comment by Yuyz0112** [comment]
@tanbowensg @xzdry I've migrated all the components to the latest slots spec.

And there are several components in the arco-lib that need check because I think it needs slot props but I do not know the usage of its slots.

- Cascader
- Tabs
- Collapse

**Comment by xzdry** [comment]
> @tanbowensg @xzdry I've migrated all the components to the latest slots spec.
> 
> And there are several components in the arco-lib that need check because I think it needs slot props but I do not know the usage of its slots.
> 
> * Cascader
> * Tabs
> * Collapse

The `tabs` are fine. `collapse` needs to be changed to the following:
```tsx
export const Collapse = implementRuntimeComponent(options)(props => {
  const { defaultActiveKey, options, ...cProps } = getComponentProps(props);
  const { elementRef, mergeState, slotsElements, customStyle, callbackMap } = props;

  const [activeKey, setActiveKey] = useState<string[]>(defaultActiveKey.map(String));

  const onChange = useCallback(
    (currentOperateKey: string, activeKey: string[]) => {
      setActiveKey(activeKey);
      mergeState({ activeKey });
      callbackMap?.onChange?.();
    },
    [callbackMap, mergeState]
  );

  return (
    <BaseCollapse
      ref={elementRef}
      className={css(customStyle?.content)}
      {...cProps}
      activeKey={activeKey}
      onChange={onChange}
    >
      {options.map((o, idx) => {
        const { key, ...props } = o;
        return (
          <BaseCollapse.Item key={key} name={String(key)} {...props}>
            {slotsElements?.content ? (
              <slotsElements.content activeKey={String(key)} />
            ) : null}
          </BaseCollapse.Item>
        );
      })}
    </BaseCollapse>
  );
});

```

As for the `Cascader` component itself there seems to be some problems, I'll check it later and change it

**Comment by Yuyz0112** [comment]
@tanbowensg 

I've optimized the render process to avoid the new slot elements object causing re-mount.

Now, the process looks like this:

render an app
  render a component
    get slot elements map for this component: the map will only change when props other than `slotProps` change.
      create `<ImplWrapper />` for slot element: it will keep the same instance if the slot elements map is not re-new.
        when slot props change, it will cause the `<ImplWrapper />` re-render.

If I understand correctly, you are trying to make the `<ImplWrapper />` static and state manager watchers should only cause all the re-render.
I like this idea, but currently, slot props break this design. Although this does not cause extra re-render, I'd love to re-implement it in the next PR by creating a slot props store and making this `<ImplWrapper />` static again.

**Comment by tanbowensg** [comment]
> Although this does not cause extra re-render
In my testing, I found this pr will cause component re-render after its siblings component changes in editor. I worry about that this may lead to poor performance in heavy app, like cluster-installer. We need to publish a beta version to see how will it affect cluster-installer.

**Comment by Yuyz0112** [comment]
> In my testing, I found this pr will cause component re-render after its siblings component changes in the editor.

This is not the expected behavior. Could you share a minimal schema to reproduce? I'm going to fix this before merging.

**Comment by tanbowensg** [comment]
If I change `input3`'s property, everything is ok. If I change `input2`'s property, both `input1` and `input2` will remount.

https://user-images.githubusercontent.com/12260952/168257862-8c0c66bf-6a45-4d76-8da4-02afb3bc49e9.mov


```
{
  "version": "sunmao/v1",
  "kind": "Application",
  "metadata": {
    "name": "some App"
  },
  "spec": {
    "components": [
      {
        "id": "box1",
        "type": "chakra_ui/v1/box",
        "properties": {
          "w": 40,
          "h": 40,
          "minW": "",
          "maxW": "",
          "minH": "",
          "maxH": "",
          "m": "",
          "p": "",
          "overflow": "",
          "bg": "",
          "bgColor": "",
          "opacity": "",
          "bgGradient": "",
          "bgClip": "",
          "bgImage": "",
          "bgSize": "",
          "bgPosition": "",
          "bgRepeat": "",
          "bgAttachment": "",
          "boxShadow": "",
          "border": "1px solid black",
          "borderRadius": ""
        },
        "traits": [
          {
            "type": "core/v1/style",
            "properties": {
              "styles": []
            }
          }
        ]
      },
      {
        "id": "input1",
        "type": "arco/v1/input",
        "properties": {
          "allowClear": false,
          "disabled": false,
          "readOnly": false,
          "defaultValue": "",
          "placeholder": "please input",
          "error": false,
          "size": "default"
        },
        "traits": [
          {
            "type": "core/v1/slot",
            "properties": {
              "container": {
                "id": "box1",
                "slot": "content"
              }
            }
          }
        ]
      },
      {
        "id": "input2",
        "type": "arco/v1/input",
        "properties": {
          "allowClear": false,
          "disabled": false,
          "readOnly": false,
          "defaultValue": "",
          "placeholder": "please input",
          "error": false,
          "size": "default"
        },
        "traits": [
          {
            "type": "core/v1/slot",
            "properties": {
              "container": {
                "id": "box1",
                "slot": "content"
              }
            }
          }
        ]
      },
      {
        "id": "input3",
        "type": "arco/v1/input",
        "properties": {
          "allowClear": false,
          "disabled": false,
          "readOnly": false,
          "defaultValue": "",
          "placeholder": "please input",
          "error": false,
          "size": "default"
        },
        "traits": []
      }
    ]
  }
}
```

**Comment by Yuyz0112** [comment]
@tanbowensg 

The re-mount issue is caused by creating react element inline when we collect slots elements. So I change the slot elements interface from the React element to a plain function with the React nodes as return values.

Now the usage of slot elements looks like this:
```
            {slotsElements?.content ? (
              slotsElements.content({ activeKey, index })
            ) : (
              <EmptyPlaceholder componentName="" />
            )}
```
and the re-mount issue has been fixed.

**Comment by Yuyz0112** [comment]
@xzdry I've updated the Collapse component's implementation.

**Comment by Yuyz0112** [comment]
still has some design needs a double-check, mark as a draft at this moment

File: packages/core/src/slot.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import { JSONSchema7 } from 'json-schema';
+
+export type SlotSchema = {
```
Comment (MrWindlike) [inline packages/core/src/slot.ts:3]: May name it `SlotSpec`?

Comment (tanbowensg) [reply packages/core/src/slot.ts]: ok

**Pull request merged**